src/tensor_parallel/slicing_configs.py

```diff
@@ -78,13 +88,76 @@ def split_num_heads(num_heads: int, *, rank: int, world_size: int):
     alibi_expanded = alibi.reshape(-1, num_heads, *alibi.shape[1:])
     alibi_part = alibi_expanded.tensor_split(world_size, dim=1)[rank]
     return alibi_part.reshape(-1, *alibi.shape[1:])
 
 
+def get_t5_config(model_config: T5Config, devices: Sequence[torch.device]) -> Config:
+    world_size = len(devices)
+    num_heads = model_config.num_heads
+    head_dim = model_config.d_kv
+
+    gather_kv_across_ranks = CollectiveOperation(
+        world_size=world_size, func=lambda *kvs: gather_kv(*kvs, world_size=world_size)
+    )  # this operation ensures that we get attention cache for all heads on each device
+
+    def select_kv_for_rank(*kvs, rank):
+        if kvs[0] is None:
+            return None
+        else:
+            kvs = kvs[0]
+            return (kvs[0][rank], kvs[1][rank])
+
+    return Config(
+        state_rules={
+            r".*SelfAttention\.q\.(weight|bias)$": partial(
+                split_heads, dim=0, head_dim=head_dim, world_size=world_size
+            ),
+            r".*SelfAttention\.k\.(weight|bias)$": partial(
+                split_heads, dim=0, head_dim=head_dim, world_size=world_size
+            ),
+            r".*SelfAttention\.v\.(weight|bias)$": partial(
+                split_heads, dim=0, head_dim=head_dim, world_size=world_size
+            ),
+            r".*relative_attention_bias\.weight$": "split 1",
+            r".*SelfAttention\.o\.weight$": partial(split_heads, dim=1, head_dim=head_dim, world_size=world_size),
+            r".*DenseReluDense\.wi\.weight$": "split 0",
+            r".*DenseReluDense\.wi_0\.weight$": "split 0",
+            r".*DenseReluDense\.wi_1\.weight$": "split 0",
+            r".*DenseReluDense\.wo\.weight$": "split 1",
+            r".*shared.weight$": "split 1",
+            r".*lm_head\.weight$": "split 1",
+            # note: ^-- lm_head.weight tied with word embeddings
+        },
+        input_rules={
+            r".*SelfAttention$": {"past_key_value": select_kv_for_rank},
+            r".*lm_head$": {0: "split -1"},  # note: we need to split lm_head inputs because
+            # ... lm_head's weights (tied embeddings) are already split across input dimension
+        },
+        output_rules={
+            r".*SelfAttention$": {0: "sum", 1: gather_kv_across_ranks},
+            r".*DenseReluDense$": {0: "sum"},
+            r".*shared$": {0: "gather -1"},
+            r".*lm_head$": {0: "sum"},
+        },
+        attr_rules={
+            r".*SelfAttention$": {
+                "n_heads": partial(split_num_heads, world_size=world_size),
+                "inner_dim": partial(split_inner_dim, num_heads=model_config.num_heads, world_size=world_size),
+            },
+            r".*relative_attention_bias$": {"embedding_dim": partial(split_num_heads, world_size=world_size)},
+        },
+    )
+
+
 PREDEFINED_CONFIGS: Dict[str, ConfigGetter] = {
     "BloomModel": get_bloom_config,
+    "T5ForConditionalGeneration": get_t5_config,
```

Comment:
quick sanity check: are we sure this should apply to T5ForConditionalGeneration and not, say, T5Model?

Reply:
There are [model architectures ](https://huggingface.co/google/flan-t5-xxl/blob/main/config.json#L3) and there is [model type](https://huggingface.co/google/flan-t5-xxl/blob/main/config.json#L15). I'm not quite sure what to use